src/generator/renderers/content.tsx

```diff
@@ -129,18 +132,25 @@ export async function MessageSingleASTNode({ node, context }: { node: SingleASTN
 
     case 'channel': {
       const id = node.id as string;
-      const channel = await context.callbacks.resolveChannel(id);
+      // NOTE: resolveChannel is now assumed to be sync or returns null
+      const channel = (
+        context.callbacks.resolveChannel as (
+          id: string
+        ) => { isDMBased?: () => boolean; name: string; type: ChannelType } | null
+      )?.(id);
 
       return (
-        <DiscordMention type={channel ? (channel.isDMBased() ? 'channel' : getChannelType(channel.type)) : 'channel'}>
-          {channel ? (channel.isDMBased() ? 'DM Channel' : channel.name) : `<#${id}>`}
+        <DiscordMention type={channel ? (channel.isDMBased?.() ? 'channel' : getChannelType(channel.type)) : 'channel'}>
+          {channel ? (channel.isDMBased?.() ? 'DM Channel' : channel.name) : `<#${id}>`}
         </DiscordMention>
       );
     }
 
     case 'role': {
       const id = node.id as string;
-      const role = await context.callbacks.resolveRole(id);
+      const role = (context.callbacks.resolveRole as (id: string) => { name: string; hexColor: string } | undefined)?.(
+        id
+      );
```

Comment:
random type cast again. We have control over the callbacks struct type, why not change it there?